u2-routing-controllers/app.js

```diff
@@ -0,0 +1,37 @@
+// import opject in file
+const express = require('express')
+const { urlencoded, json } = require('body-parser')
+const { getUserByQuery, getUserByPath, getUserByForm, search } = require('./features/userController')
+const { logger } = require('./_services/logger')
+const { index } = require('./features/indexController')
+const { getproduct } = require('./features/productController')
+const { getrgb } = require('./features/rgbController')
+const { gettemp } = require('./features/tempController')
+const { getecho, getechoByform } = require('./features/echoController')
+
+const app = express()
+
+// middlewares โค๊ดที่ใช้
+app.use(urlencoded({ extended: false }))
+app.use(json()) // json() = function that creates middleware
+app.use(logger) // logger = a middleware
+
+// Routes เชื่อมไปยังตัวควบคุม
+// index lab
+app.get('/', index)
+// User
+app.get('/users', getUserByQuery)
+app.get('/users/search', search)
+app.get('/users/:username', getUserByPath)
+app.post('/users', getUserByForm)
+// Lab
+app.get('/products/search', getproduct)
+app.get('/rgb', getrgb)
+app.get('/:u1-to-:u2/:t1', gettemp)
+app.get('/echo', getecho)
+app.post('/echo', getechoByform)
```

Comment:
ดี! สังเกต 3 กรณี:
- **req.query.xxx -- query string ที่ไม่ได้ประกาศใน `app.js`**  
  `/products/search?q=abc&page=4`  
  `/rgb?r=255&g=128&b=0`
- **req.params.xxx -- path parameters ประกาศใน `app.js` โดยใช้ `:`**  
  `/:u1-to-:u2/:t1`  
- **req.body.xxx -- ข้อมูลถูกส่งในเนื้อหาของคำขอ POST**  
  GET `/echo` ซึ่งให้บริการแบบฟอร์ม; แบบฟอร์มกำหนดชื่อของตัวแปร  
  POST `/echo` ซึ่งรับข้อมูลที่โพสต์